File: CPP/ByRound/2115/2115_D_Gellyfish_and_Forget_Me_Not.cpp
/*
 * Problem URL : https://codeforces.com/problemset/problem/2115/D
 * Submit Date : 2025-08-16
 */

#include <bits/stdc++.h>
using namespace std;
using ll = long long;
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);
  int t;
  cin >> t;
  while (t--) {
    int n;
    vector<ll> a;
    ll Bas = 0;
    string s;
    cin >> n;
    a.resize(n);
    for (auto &k : a)
      cin >> k, Bas ^= k;
    for (auto &k : a) {
      ll x;
      cin >> x;
      k ^= x;
    }
    cin >> s;
    for (int z = 59; z >= 0; z--) {
      int p = -1;
      for (int i = n - 1; i >= 0; i--)
        if (a[i] >> z & 1) {
          if (p == -1)
            p = i;
          else
            a[i] ^= a[p];
        }
      if (p == -1)
        continue;
      if (s[p] == '0')
        Bas = min(Bas, Bas ^ a[p]);
      else
        Bas = max(Bas, Bas ^ a[p]);
      a[p] = 0;
    }
    cout << Bas << '\n';
  }
}

File: Python/ByRound/2115/2115_D_Gellyfish_and_Forget_Me_Not.py
# Contest 2115, Problem D: Gellyfish and Forget-Me-Not
# URL: https://codeforces.com/contest/2115/problem/D

import sys

input = sys.stdin.readline

from math import gcd

out = []

t = int(input())
for _ in range(t):
    n = int(input())
    a = list(map(int, input().split()))
    b = list(map(int, input().split()))
    c = input().strip()

    base = 0
    for i in range(n):
        base ^= a[i]
        b[i] ^= a[i]

    fixed = [0] * n
    for bt in range(60)[::-1]:
        mask = 1 << bt

        i = n - 1
        while i >= 0:
            if b[i] & mask == 0 or fixed[i]:
                i -= 1
                continue

            diff = b[i]
            fixed[i] = 1

            if c[i] == "1":
                base = max(base, base ^ diff)
            else:
                base = min(base, base ^ diff)

            break

        # print(bt, i, b)

        i -= 1
        while i >= 0:
            if b[i] & mask:
                b[i] ^= diff
            i -= 1

    out.append(base)

print("\n".join(map(str, out)))
